### ISSUE TITLE
moveit_setup_assistant crashing on URDF load again

### ISSUE THREAD

**Issue by HWiese1980** [moveit/moveit_setup_assistant#60]
This is the URDF it crashes on: http://pastebin.com/9E0CXQ0Q
And here's the back trace: http://pastebin.com/rmF9AWwZ

I've run a complete apt-get upgrade across all my ROS Hydro packages before trying it again. So everything should be todays most current version. Still, it's crashing.

**Comment by isucan** [comment]
Are you running a 32bit os?

**Comment by HWiese1980** [comment]
Yeah, I do. Is that still an issue? Do you know of a workaround?

**Comment by isucan** [comment]
The error you see seems to be an alignment issue when using a map of Eigen transforms. The issue does not come up on 64bit OS. The code itself specifies the use of an Eigen-specific allocator for the map, so I am not sure how to correct this in code. Please let me know if you find a solution.

**Comment by HWiese1980** [comment]
So, it is a bug that is kinda out of your reach?

**Comment by isucan** [comment]
I guess so. If anyone knows how to fix it, I would appreciate it.

**Comment by hersh** [comment]
I know a little about this, I fixed it once somewhere else.

Here is the main Eigen page on it: http://eigen.tuxfamily.org/dox-devel/group__TopicUnalignedArrayAssert.html

There is also this: http://eigen.tuxfamily.org/bz/show_bug.cgi?id=490 which would be caused outside of our code entirely.  What gcc version was used to build it?  That bug happens on gcc 4.7.

I'll look into the moveit code and see if I can help any more.

**Comment by isucan** [comment]
http://eigen.tuxfamily.org/dox-devel/group__TopicUnalignedArrayAssert.html should be in theory addressed. I did not know about the gcc problem though.

**Comment by hersh** [comment]
Hmm, the moveit code looks like it does what the Eigen docs recommend.  However looking at http://eigen.tuxfamily.org/dox-devel/Memory_8h_source.html the source for Eigen::aligned_allocator, I don't see how this would make pair<void*,EigenStruct> work any better.  Seems to me it would align the beginning of the pair, but not the second part which is what needs to be aligned.

**Comment by hersh** [comment]
Ah, I see... http://eigen.tuxfamily.org/dox-devel/group__TopicStructHavingEigenMembers.html this explains it.  There's an attribute on the Eigen structs which makes them be aligned _within a struct_ to the right alignment.  So then the only thing you need is to align the struct itself.

Since MoveIt is already doing this, I'm thinking @SeveQ is using gcc 4.7 and is hitting that bug, which really is out of our hands.

If that is the problem, a workaround is to use gcc 4.6.

**Comment by HWiese1980** [comment]
I'm using the precompiled moveit binaries that come with the ros-hydro-moveit-\* Ubuntu packages provided by packages.ros.org. The only binaries that I compile myself are self-made ROS packages.

**Comment by hersh** [comment]
@SeveQ Which version of Ubuntu are you running?

**Comment by HWiese1980** [comment]
Kubuntu based on...

bash# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=13.04
DISTRIB_CODENAME=raring
DISTRIB_DESCRIPTION="Ubuntu 13.04"

**Comment by hersh** [comment]
http://packages.ubuntu.com/search?suite=raring&section=all&arch=any&keywords=gcc&searchon=names

Yup, looks like Ubuntu 13.04 uses gcc 4.7, which has this issue.

So the workaround I know of is to install gcc 4.6 and build moveit from source.

It is probably worth looking into the bug in libstdc++ mentioned here: http://gcc.gnu.org/bugzilla/show_bug.cgi?id=53900 to see if any progress has been made.

**Comment by hersh** [comment]
Looks here: http://eigen.tuxfamily.org/index.php?title=ChangeLog like Eigen version 3.1.3 may have a fix for this issue with gcc 4.7.

Raring (Ubuntu 13.04) seems to use Eigen version 3.1.2: http://packages.ubuntu.com/search?suite=raring&section=all&arch=any&keywords=libeigen3-dev&searchon=names

Maybe the maintainer for the ubuntu Eigen package could be convinced to update to 3.1.3?

**Comment by HWiese1980** [comment]
Well, I don't really like the thought of installing another GCC and building moveit completely from sources. There'll probably just be another string of issues I'll have to deal with... I'm only a small ROS developer, no kernel/lib hacker. 

Alright, I think I'll try to install Eigen 3.1.3 from sources which should, at least for me, hopefully, be a far easier approach than installing another GCC. I'll keep you informed if it helps...

**Comment by hersh** [comment]
For reference, here's the bug report I entered against the ubuntu package of Eigen: https://bugs.launchpad.net/ubuntu/+source/eigen3/+bug/1240221

**Comment by HWiese1980** [comment]
Great! Thanks! 

I failed to compile Eigen 3.1.4 (which is latest 3.1 stable) from source... now I'm eagerly waiting for a response from the Eigen maintainer.

**Comment by HWiese1980** [comment]
Alright, as I expected he can't lift Eigen to >= 3.1.3 in Raring since Raring is a stable release. He recommends upgrading my system to Saucy. I'm too scared of doing that. It says: never change a running system! I have had severe difficulties getting my system running the way it does due to some driver issues. I don't want to go through that again... any other idea maybe on how to solve this problem aside from messing around with my system internals?

**Comment by beetleskin** [comment]
I "built" Eigen 3.1.4 (as there are only Headers, you don't have to compile anything as I understand it), make installed it to /usr/... and built MoveIt from source. Still get the very same Segfault. I'm on raring 32bit with gcc 4.7 and hydro.

**Comment by HWiese1980** [comment]
Since, as you mentioned, there are only headers, there most likely won't be any functional difference between 3.1.2 and 3.1.3/3.1.4 and hence no bug fixed.

**Comment by hersh** [comment]
@SeveQ have you looked at the headers in question?  They are monsters.  All the functionality of the Eigen "library" is in header files.

I'm very disappointed to hear that 3.1.4 doesn't fix it.  Though this perhaps leaves open the possibility that it can be fixed in MoveIt code instead.

**Comment by HWiese1980** [comment]
Why, that's outrageous! Emh, no... to be honest. But makes sense. At least it explains how such a bug could slip through. I'd call that a severe design flaw. Aren't there any alternative libraries providing Eigen's functionality?

**Comment by mikeferguson** [comment]
@SeveQ Eigen is a template library, hence all code is in headers. By doing that way, it ensures fast, customized code, for any size matrix.

Honestly, we use Eigen all over ROS, and there have been a number of issues with Eigen on 32-bit systems. Is there any particular reason you need 32-bit? Or would it be possible to use a 64-bit system?

**Comment by HWiese1980** [comment]
I've been told never to put actual code into a header file. Well, seems that announcement was kind of relative... The particular reason I was stuck to 32bit was my graphics driver. I experienced serious difficulties installing my system because of some misbehavior of the driver of which I could never figure out the cause. After installing the newest vendor driver my system wouldn't boot anymore and just greeted me with a bucket full of graphical glitches. Anyway, I got it running after a long time of hassle and didn't want to go through that again. Not to mention my concerns wether the driver that I found working on 32bit after all would work on 64bit as well. 

TL;DR: Had trouble with my graphics driver, after weeks I found one that worked. Now, I've made an effort and switched to 64bit, the graphics driver's fortunately working on 64bits as well. So the bug isn't an issue for me anymore. At least. 

Still something that, in my opinion, must be solved sooner or later.

**Comment by mikeferguson** [comment]
@SeveQ From looking at the tickets, should be fixed in Saucy with latest Eigen patches, but won't be fixed in Raring (and that is an Ubuntu decision, not something we can really fix).

Is there a particular reason you are using the non-LTS version of Ubuntu? Personally, I've always found that the non-LTS versions tend to be much less stable (especially the one right before a new LTS), and with the new 9-mo support cycle for non-LTS releases, it seems that will be an even more likely event.

**Comment by HWiese1980** [comment]
@mikeferguson Good question! Actually, there is no particular reason I'm using a non-LTS Ubuntu. Doesn't the bug exist in 12.04 as well? I'm not that familiar with the support plans of Ubuntu. Would LTS imply that a bug like this still had to be fixed although the whole release is already considered stable?

**Comment by mikeferguson** [comment]
@SeveQ Thing thing about an LTS release, is that they tend not to change lots of things right before finalizing it. Thus, you can sort of push all sorts of terrible code into 13.10 (in fact, the odd.10 release is usually a mess because everybody rushes things into it so that they can get testing for the even.04 release which is LTS), but in theory, 14.04 will tend to much more tested/stable.

As for this particular problem, 12.04 does not use GCC 4.7, so it should not be a problem, and the next LTS (14.04) should probably be OK since Eigen and GCC will have learned to play better together (hopefully).

I was about to make a comment that hopefully at some point, Ubuntu will also default to 64-bit for desktop (they already do for server), and all these Eigen alignment issues pretty much disappear -- but just looked to see and in fact the download pages now default to 64-bit. I know this discussion came up before with Point Cloud Library, as they had a number of 32-bit issues and were talking about dropping support for 32-bit altogether (at that point, Ubuntu still recommended 32-bit for desktop).

**Comment by HWiese1980** [comment]
@mikeferguson I see. For now, after switching to 64 bits, everything relevant to me appears to be fine. So, I guess I'll leave it that way until I run into another release related issue and think about installing 12.04 64 bits then. If I remember correctly the graphics driver issues occurred in 12.04 as well (had it running several months ago, don't know anymore why I upgraded). So I'd have to go through that again. Something I'd like to avoid for as long as possible. Looks like I'll need a new notebook sooner or later...

Thanks for the insight!

**Comment by beetleskin** [comment]
I'm going to install 12.4 64bit then. hurrayyyy :) As for me you might close this issue, since it's not your bug after all. 

Bottomline: There is a bug with GCC 4.7 on Ubuntu 13.4 32bit with Eigen 3.1.2. It's not likely to being fixed, so upgrade/downgrade your system to 13.4 64bit resp. 12.4. 
Eigen 3.1.4 together with MoveIt built from source did not help.

**Comment by davetcoleman** [comment]
I've noted this on the installation
pagehttp://moveit.ros.org/wiki/Installation#Installation_and_Setup

**Comment by UltronDestroyer** [comment]
I'm getting this error as well, except I have 12.04, 64bit OS, and gcc 4.6.3.

**Comment by davetcoleman** [comment]
Solution was found:

> Bottomline: There is a bug with GCC 4.7 on Ubuntu 13.4 32bit with Eigen 3.1.2. It's not likely to being fixed, so upgrade/downgrade your system to 13.4 64bit resp. 12.4. 

Closing.
